### PR TITLE
add some precompile for Artifacts and add a type annotation

### PR DESCRIPTION
**Pull request by KristofferC**
Showed up quite a bit in some package load times.

### CODE REVIEW

File: contrib/generate_precompile.jl [hunk 1]
```diff
@@ -141,7 +141,10 @@ if Artifacts !== nothing
     precompile_script *= """
     using Artifacts, Base.BinaryPlatforms, Libdl
     artifacts_toml = abspath(joinpath(Sys.STDLIB, "Artifacts", "test", "Artifacts.toml"))
-    # cd(() -> (name = "HelloWorldC"; @artifact_str(name)), dirname(artifacts_toml))
```
Comment (KristofferC) [inline contrib/generate_precompile.jl:144]: One reason this didn't work before is that the macro is expanded before the `cd` happens. But even so, it won't find `HelloWorldC` as an artifact. So I changed to `macroexpand`.

File: stdlib/Artifacts/src/Artifacts.jl [hunk 2]
```diff
@@ -418,7 +418,7 @@ collapsed artifact.  Returns `nothing` if no mapping can be found.
 """
 function artifact_hash(name::String, artifacts_toml::String;
                        platform::AbstractPlatform = HostPlatform(),
-                       pkg_uuid::Union{Base.UUID,Nothing}=nothing)
+                       pkg_uuid::Union{Base.UUID,Nothing}=nothing)::SHA1
```
Comment (simeonschaub) [inline stdlib/Artifacts/src/Artifacts.jl:421]: This can also return `nothing`, no?

Comment (KristofferC) [reply stdlib/Artifacts/src/Artifacts.jl:421]: Ouch... I added that as an afterthought just before pushing. Let's see if the test catch it ;)

Comment (KristofferC) [reply stdlib/Artifacts/src/Artifacts.jl:421]: It didn't..

**Review by staticfloat — APPROVED** [review]

**Pull request merged**